queue.go

```diff
@@ -241,173 +251,152 @@ func (queue *redisQueue) StopConsuming() <-chan struct{} {
-	return count > 0
-}
-
-func (queue *redisQueue) addConsumer(tag string) string {
-	if queue.deliveryChan == nil {
-		log.Panicf("rmq queue failed to add consumer, call StartConsuming first! %s", queue)
-	}
-
-	name := fmt.Sprintf("%s-%s", tag, uniuri.NewLen(6))
-
-	// add consumer to list of consumers of this queue
-	if ok := queue.redisClient.SAdd(queue.consumersKey, name); !ok {
-		log.Panicf("rmq queue failed to add consumer %s %s", queue, tag)
+	name, err = queue.addConsumer(tag)
+	if err != nil {
+		return "", err
 	}
-
-	// log.Printf("rmq queue added consumer %s %s", queue, name)
-	return name
-}
-
-func (queue *redisQueue) RemoveAllConsumers() int {
-	count, _ := queue.redisClient.Del(queue.consumersKey)
-	return count
+	go queue.consumerConsume(consumer)
+	return name, nil
 }
 
-func (queue *redisQueue) consume() {
+func (queue *redisQueue) consumerConsume(consumer Consumer) {
+	defer queue.stopWg.Done()
 	for {
-		batchSize := queue.batchSize()
-		wantMore := queue.consumeBatch(batchSize)
-
-		if !wantMore {
-			time.Sleep(queue.pollDuration)
-		}
-
-		if atomic.LoadInt32(&queue.consumingStopped) == int32(1) {
-			// log.Printf("rmq queue stopped consuming %s", queue)
-			close(queue.deliveryChan)
-			// log.Printf("rmq queue stopped fetching %s", queue)
+		select {
+		case <-queue.consumingStopped: // prefer this case
 			return
+		default:
 		}
-	}
-}
-
-func (queue *redisQueue) batchSize() int {
-	// prefetchCount := len(queue.deliveryChan)
-	// prefetchLimit := queue.prefetchLimit - prefetchCount
-	prefetchLimit := queue.prefetchLimit - queue.UnackedCount() 
 
-	// TODO: ignore ready count here and just return prefetchLimit?
-	if readyCount := queue.ReadyCount(); readyCount < prefetchLimit {
-		return readyCount
-	}
-	return prefetchLimit
-}
+		select {
+		case <-queue.consumingStopped:
+			return
 
-// consumeBatch tries to read batchSize deliveries, returns true if any and all were consumed
-func (queue *redisQueue) consumeBatch(batchSize int) bool {
-	if batchSize == 0 {
-		return false
-	}
+		case delivery, ok := <-queue.deliveryChan:
+			if !ok { // deliveryChan closed
+				return
+			}
 
-	for i := 0; i < batchSize; i++ {
-		value, ok := queue.redisClient.RPopLPush(queue.readyKey, queue.unackedKey)
-		if !ok {
-			// debug(fmt.Sprintf("rmq queue consumed last batch %s %d", queue, i)) // COMMENTOUT
-			return false
+			consumer.Consume(delivery)
 		}
-
-		// debug(fmt.Sprintf("consume %d/%d %s %s", i, batchSize, value, queue)) // COMMENTOUT
-		queue.deliveryChan <- newDelivery(value, queue.unackedKey, queue.rejectedKey, queue.pushKey, queue.redisClient)
 	}
+}
 
-	// debug(fmt.Sprintf("rmq queue consumed batch %s %d", queue, batchSize)) // COMMENTOUT
-	return true
+// AddConsumerFunc adds a consumer which is defined only by a function. This is
+// similar to http.HandlerFunc and useful if your consumers don't need any
+// state.
+func (queue *redisQueue) AddConsumerFunc(tag string, consumerFunc ConsumerFunc) (string, error) {
```

Comment:
Not quite, even though it looks like that should work.

This works:
```go
queue2.AddConsumerFunc("cons-func", func(delivery Delivery) {})
```
(because by passing that anonymous func into `AddConsumerFunc` it gets silently converted to be a `ConsumerFunc` which implements the `Consumer` interface.

This doesn't:
```go
queue2.AddConsumer("cons-func", func(delivery Delivery) {})
```
(because `func()` itself doesn't support the consumer interface)

This would work:
```go
queue2.AddConsumer("cons-func", ConsumerFunc(func(delivery Delivery) {}))
```
(because here we manually type the func to support the consumer interface)

So in a way the additional function is a shorthand for the third version above.
 (It's very similar to `http.Handle()` and `http.HandleFunc()` btw)